If there is a previous backup to base the time off we still need to s…

joshua-rule opened the pull request:
…et the backup hour

 so that the backup time can be changed

[PM-2995]

# About this change: What it does, why it matters

(all contributors please complete this section, including maintainers)


[PM-2995]: https://aiven.atlassian.net/browse/PM-2995?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

aiven-amartin commented on myhoard/controller.py:
```diff
@@ -880,7 +880,8 @@ def _current_normalized_backup_timestamp(self) -> str:
                 normalized = normalized - datetime.timedelta(days=1)
-            normalized = normalized.replace(hour=backup_hour, minute=backup_minute, second=0, microsecond=0)
+        
+        normalized = normalized.replace(hour=backup_hour, minute=backup_minute, second=0, microsecond=0)
```
I think you need to pull the whole if statement out of the else, as we need to do the date math to make sure our window isn't > 24h.

aiven-amartin replied:
It also would be worth excavating if the actual bug is just that `_previous_normalized_backup_timestamp` parses all streams, rather than just active basebackup ones. I.e. I expect the reason we check the old one is so we don't change the request time while a stream is active, but then we get stuck in this perma-loop. I think we should just remove the state checking behaviour, just would need to validate how things behave when we change this in the middle of a backup (would not want to abort).

joshua-rule replied:
It needs to check the previous backup time because if the backup interval is greater than 24 hours the previous logic will still run the backup at least once a day.

I started trying to solve the problem more generally but loads of problems arise because the interval is in minutes so the time drifts every day making the backup hour/minute number kind of meaningless.

I think maybe the simplest but not so nice way to fix our immediate problem with this is to check if the interval is greater than 24 hours and use the current logic to allow intervals greater than 24 hours, otherwise use the old logic so you can change the backup time.

Otherwise if we enforced some kind of allowed values on the interval, like hourly then if you get past 24 hours you can only have values that have 24 hours as a factor this all becomes a lot more sensible.

aiven-amartin replied:
Makes sense to me. Can just update the documentation to say for intervals != 24h, the backup hour/minute configuration is not respected, as even if its less than 24hrs it doesn't really make sense either.

aiven-amartin replied:
Ideally we would reject configuration that sets both sets to non-defaults but that would be a breaking change now I guess.

aiven-amartin commented on myhoard/controller.py:
```diff
@@ -875,7 +875,7 @@ def _current_normalized_backup_timestamp(self) -> str:
         previous_normalized = self._previous_normalized_backup_timestamp()
         # in case we have a previous backup time, we can use it as starting point
-        if previous_normalized:
+        if previous_normalized and backup_interval_minutes > 1440:
```
```suggestion
        if previous_normalized and backup_interval_minutes != 1440:
```

And this magic should probably be a magic value from somewhere, but that can be a follow up I guess. Also `>` is good enough for now I think.

joshua-rule committed:
If the backup interval is in days (1 day, 2 days etc.) allow changing the time of day backups are taken.
If the backup interval is anything else then we use the backup hour/minute setting for the first backup,
afterwards the backup time is based solely of the previous backup time and cannot be changed.

aiven-amartin commented on README.md:
```diff
@@ -189,15 +189,16 @@ than 24 hours this is used as base for calculating the backup times. E.g. if
-**backup_minute**
-
-The minute of hour at which to take new full backup.
+Note: If this value is set to > 1440 (1 day) then the backup_hour and backup_minute
```
```suggestion
Note: If this value is set to anything other than 1440 (1 day) then the backup_hour and backup_minute
```

aiven-amartin commented on README.md:
```diff
@@ -189,15 +189,16 @@ than 24 hours this is used as base for calculating the backup times. E.g. if
-
-The minute of hour at which to take new full backup.
+Note: If this value is set to > 1440 (1 day) then the backup_hour and backup_minute
+settings cannot be changed once the first backup has been taken
```
```suggestion
settings cannot be changed once the first backup has been taken, as having a cycle != 1 day
means that the hour and minute of the backup will not be the same each day.
```

joshua-rule commented on README.md:
```diff
@@ -189,15 +189,17 @@ than 24 hours this is used as base for calculating the backup times. E.g. if
-**backup_minute**
-
-The minute of hour at which to take new full backup.
+Note: If this value is set to anything other than 1440 (1 day) then the backup_hour and backup_minute
```
yea we can allow multiples now fairly easily, have made that change

joshua-rule commented:
looks like older versions of xtrabackup have disappeared from the repo, didn't get a chance to look into why that is.

kmichel-aiven commented on test/test_controller.py:
```diff
@@ -1357,6 +1357,71 @@ def flush_binlogs():
+
+    # After second backup, the next scheduled one should be at 2023-02-04 03:00:00, but let's change
+    # backup_interval_minutes to 2880 (48 hours)
+    # m_controller.backup_settings["backup_interval_minutes"] = 2880
```
I don't think this line should be commented ?

Given the comment, it should be there, and if it shouldn't, then it should be fully removed.

kmichel-aiven commented:
> looks like older versions of xtrabackup have disappeared from the repo, didn't get a chance to look into why that is.

That's not ideal :(

Well, since this PR is purely internal logic, as long as the 8.0.30 tests pass, I'm fine with merging it, we can solve the repo problem in a separate PR.
